dramatic performance boost for consecutive rospack calls [backport to indigo]

rhaschke opened the pull request:
Consecutive calls to the rospack library, e.g. `ros::package::getPath()`, from within the same process recrawled the `ROS_PACKAGE_PATH` all the time, because the appropriate rospack instances were created again and again.
This PR (going in line with a similar PR on roslib) improves things in two respects:
- a static `rospack::Rospack` instance is used in `ROSPack::run()`, thus re-using the same instance for all consecutive calls
- in `Rosstackage::crawl()` the cache is loaded only once at the beginning, when `search_paths_` is still empty

The following simple test program (doing 3000 calls to `getPath()`) performs 160 times faster:

```
time ./foo
real    0m0.264s
user    0m0.251s
sys     0m0.012s

time ./foo
real    0m49.456s
user    0m42.500s
sys     0m6.840s
```

``` cpp
#include <ros/package.h>
int main(int argc, char *argv[]) {
    for (int i=0; i<3000; ++i) {
        ros::package::getPath("roscpp");
    }
}
```

rhaschke committed:
allow caching of rospack results

rhaschke committed:
rospack: read cache file only once in a process

rhaschke commented:
This goes along with [roslib PR #97](https://github.com/ros/ros/pull/97).

dirk-thomas commented on src/rospack.cpp:
```diff
@@ -331,16 +331,13 @@ Rosstackage::crawl(std::vector<std::string> search_path,
   {
-    if(readCache())
+    // read the cache only once (when search_paths_ is still empty)
+    if(search_paths_.empty() && readCache())
```
Why do you want to prevent using the existing cache here? This will result in worse performance when this function is called multiple times.

rhaschke replied:
Indeed, if the current search_path has changed, we should read the corresponding cache. I adapted the code, such that there are the following code paths in Rosstackage::crawl() now:
- crawling is enforced -> just (re)crawl
- current search_path matches previous one and rospack already crawled: nothing todo [new]
- current search_path diverges from previous one: read corresponding cache
  - on success: nothing else todo
  - on failure (no cache or too old): re-crawl
- on re-crawling, set the search_paths_ appropriately

dirk-thomas commented on src/rospack_backcompat.cpp:
```diff
@@ -40,7 +40,8 @@ namespace rospack
 {
-  rospack::Rospack rp;
+  // allow caching of results between calls (of same process)
+  static rospack::Rospack rp;
```
This change has the chance to change the current behavior. How can you be sure that the caller want's to reuse the previous instance? Eventually the caller relies on the "new" crawl to have a different result.

Therefore similar changes have added the option to pass in a rospack instance instead, e.g. ros/ros_comm#682.

rhaschke replied:
As far as I could see, the rospack re-crawls every time when the `search_path` [fetched from the environment](https://github.com/rhaschke/rospack/blob/e1bcabef5955bb7377474ff5c45a5b9bedaaa104/src/rospack_cmdline.cpp#L195) has [changed](https://github.com/rhaschke/rospack/blob/c6dc1b4914886f135c69c5ef9eb6d08243097a5c/src/rospack.cpp#L344-L363). Hence, there is no need to pass in a new rospack instance, isn't it?

If I'm wrong, where would you pass in the rospack instance? In [roslib](https://github.com/rhaschke/ros/blob/8a12cd02bfa7ec8b0560339a3c1e0c455b763a66/core/roslib/src/package.cpp#L47) and all corresponding functions? That would break the API, which I tried to avoid.

dirk-thomas replied:
If API or ABI compatibility is desired this can be done by adding an optional argument to the function.

For API compatibility:

```
int ROSPack::run(int argc, char** argv, rospack::Rospack * rp = NULL)
```

For ABI compatibility it needs to be a separate function:

```
int ROSPack::run(int argc, char** argv) {
  rospack::Rospack rp;
  return run(argc, argv, &rp);
}
int ROSPack::run(int argc, char** argv, rospack::Rospack * rp)
```

While this change looks like a reasonable change since we can't rule out that some code changes the environment in between two calls this change would modify the behavior. I will bring this up in the next meeting to determine if we want to go forward with a change like this and in for which ROS distro.

rhaschke replied:
@dirk-thomas 
Thanks for discussing the issue. Please keep in mind during your discussions, that this PR aims for a caching behavior on existing code, primarily calls to [`ros::package::getPath()`](https://github.com/ros/ros/blob/jade-devel/core/roslib/src/package.cpp#L89). These calls go through [`roslib`](https://github.com/ros/ros/blob/jade-devel/core/roslib/src/package.cpp#L51) and thus don't provide the possibility to pass in a `ROSPack` instance.

What do you think of [my previous comment](https://github.com/ros/rospack/pull/49#discussion_r45936547)? Investigating the `getenv` calls in `rospack`, one can see, that the environment is evaluated for retrieving the `ROS_PACKAGE_PATH`, the cache path, and the cache timeout. As rospack re-crawls if `ROS_PACKAGE_PATH` has changed in the environment, I don't see the need to pass in a new rospack instance at all.

```
rospack.cpp:259:  char* rpp = getenv("ROS_PACKAGE_PATH");
rospack.cpp:1882:  char* ros_home = getenv("ROS_HOME");
rospack.cpp:1891:    char* home_drive = getenv("HOMEDRIVE");
rospack.cpp:1892:    char* home_path = getenv("HOMEPATH");
rospack.cpp:1902:      home_path = getenv("HOME");
rospack.cpp:1929:  char* rpp = getenv("ROS_PACKAGE_PATH");
rospack.cpp:2036:        char *rpp = getenv("ROS_PACKAGE_PATH");
rospack.cpp:2061:  const char *user_cache_time_str = getenv("ROS_CACHE_TIMEOUT");
rospack.cpp:2083:  const char* ros_package_path = getenv("ROS_PACKAGE_PATH");
```

rhaschke replied:
@dirk-thomas Did you already had a chance to discuss how to proceed on the rospack caching?

rhaschke committed:
moved code to compare search paths to new function sameSearchPaths()

readCache if search paths have changed

rhaschke commented:
@dirk-thomas Did you already have seen my answers to your remarks?

dirk-thomas commented on src/rospack.cpp:
```diff
@@ -325,45 +325,40 @@ Rosstackage::isStackage(const std::string& path)
+      return false;
+  }
+  return true;
+}
```
How is this function different from the comparison operator of a vector (http://en.cppreference.com/w/cpp/container/vector/operator_cmp)?

rhaschke replied:
Valid point. I didn't checked for availability of `vector::operator==`. Simply moved the existing verbose code into its own function. Now, the function is replaced by `std::vector::operator==`.

rhaschke committed:
replaced sameSearchPaths() by std::vector::operator==()

gerkey committed:
adding a test cache reuse

gerkey committed:
add test to check cache behavior when env vars change

gerkey commented:
I've started working on this in two branches:

https://github.com/ros/ros/tree/reuse_cache
https://github.com/ros/rospack/tree/reuse_cache

I applied @rhaschke's changes and added a test to rospack to test the effect of in-process changes to `ROS_PACKAGE_PATH`: https://github.com/ros/rospack/blob/reuse_cache/test/test/utest.cpp#L129

Sadly, that test passes on the first run, but fails on the second run and all subsequent runs. I'm still looking into it.

rhaschke commented:
Hm. I tested that only manually. I will have a look into your branch too.

rhaschke committed:
code simplification

rhaschke committed:
fixed utest

rhaschke commented:
@gerkey Created pull request #52 to fix the unit test.

rhaschke commented:
I had some issues to compile the unit test. Using my catkin tools environment, the binary rospack-utest wasn't created by default. I had to manually `make rospack-utest`. Is that normal behavior?

gerkey commented:
Thanks for fixes in #52.

But I still have failures elsewhere after the initial run, apparently related to whether the cache is available:

```
gerkey@rossum:~/code/rospack-caching/build/rospack/test$ rm -rf ~/.ros/rospack_cache_*
gerkey@rossum:~/code/rospack-caching/build/rospack/test$ ../../../devel/lib/rospack/rospack-utest 
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from rospack
[ RUN      ] rospack.reentrant
[       OK ] rospack.reentrant (4 ms)
[ RUN      ] rospack.multiple_rospack_objects
[       OK ] rospack.multiple_rospack_objects (6 ms)
[ RUN      ] rospack.deduplicate_tokens
[       OK ] rospack.deduplicate_tokens (0 ms)
[ RUN      ] rospack.env_change
[       OK ] rospack.env_change (2 ms)
[----------] 4 tests from rospack (12 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (12 ms total)
[  PASSED  ] 4 tests.
gerkey@rossum:~/code/rospack-caching/build/rospack/test$ ../../../devel/lib/rospack/rospack-utest 
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from rospack
[ RUN      ] rospack.reentrant
[       OK ] rospack.reentrant (4 ms)
[ RUN      ] rospack.multiple_rospack_objects
[       OK ] rospack.multiple_rospack_objects (5 ms)
[ RUN      ] rospack.deduplicate_tokens
[       OK ] rospack.deduplicate_tokens (0 ms)
[ RUN      ] rospack.env_change
/home/gerkey/code/rospack-caching/src/rospack/test/test/utest.cpp:188: Failure
Value of: test_pkgs.size()
  Actual: 3
Expected: output_list.size()
Which is: 4
/home/gerkey/code/rospack-caching/src/rospack/test/test/utest.cpp:204: Failure
Value of: found
  Actual: false
Expected: true
/home/gerkey/code/rospack-caching/src/rospack/test/test/utest.cpp:214: Failure
Value of: std::string()
  Actual: ""
Expected: output
Which is: "precedence1\nprecedence2\nprecedence3\nroslang"
[  FAILED  ] rospack.env_change (2 ms)
[----------] 4 tests from rospack (11 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (11 ms total)
[  PASSED  ] 3 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] rospack.env_change

 1 FAILED TEST
```

gerkey commented:
To build and run the test (assuming you have a workspace `ws` with `ros` and `rospack`):

```
cd ws
catkin_make tests
# The test is sensitive to CWD
cd build/rospack/test
../../../devel/lib/rospack/rospack-utest
```

You can also run all the tests via `catkin_make run_tests`, but you have do some scrolling to see the output of this particular test.

gerkey commented:
Here's more detail, with some instrumentation added to the test:

```
[ RUN      ] rospack.env_change
/home/gerkey/code/rospack-caching/src/rospack/test/test/utest.cpp:188: Failure
Value of: test_pkgs.size()
  Actual: 3
Expected: output_list.size()
Which is: 4
output_list: 
  precedence1
  precedence2
  precedence3
  roslang
test_pkgs: 
  precedence1
  precedence2
  precedence3
```

Looks like the result from the first run, with one setting for `ROS_PACKAGE_PATH`, is persisting into the second run, where that var has been changed.

@rhaschke, can you look into this and get back when it's working?

rhaschke commented:
Yes, of course. I will do it.

rhaschke committed:
fixed readCache()

readCache should clear stackages_ before adding more
added new method clearStackages(), used in three locations

rhaschke commented:
readCache() simply add more stackages without clearing the list of `stackages_` beforehand...
Fixed in #53.

gerkey commented:
Thanks for the fixes in #53. I'll open a new PR now.

gerkey committed:
add helpful comments

rhaschke commented:
The cache itself was valid. However, reading the cache simply added new 
stackages withouth wiping out existing stuff before.
This worked for the first run, before the internal list was empty 
anyway. Additionally, the cache invalidates after 60s thus enforcing a 
recrawl, which also wiped out internal storage beforehand.

rhaschke commented:
I rebased onto the same indigo-devel commit from where #54 started. Subsequently I rebased all additional commits in #54 onto this original PR to keep proper history (see [comments here](https://github.com/ros/ros/pull/98#issuecomment-163530021)). 
#54 and #49 are identical again (except for a [permission change](https://github.com/ros/rospack/commit/d6fa1d84429dbb298a91ba2e4cafc3f00da0cdbc) that was lost in #54 for some reason).
Hence, I suggest to close #54 and continue here again.

rhaschke commented:
@dirk-thomas, @gerkey
friendly ping on this PR (and its counter-part https://github.com/ros/ros/pull/97)

gerkey commented:
We discussed this today and agree that it should be merged into jade, then depending on whether there are any problems, we'll consider backporting to indigo.

dirk-thomas commented:
@ros-pull-request-builder test this please

dirk-thomas commented:
This patch has been merged into the newly created `jade-devel` branch: 2dbf9d8d15eafa185e99ce36777bc183f3c7b08a

If there are no regressions it will be considered for being backported into `indigo-devel`.

dirk-thomas commented:
I have updated the issue title that only the backport is pending.

rhaschke committed:
merge branch 'origin/indigo-devel' into indigo-devel

dirk-thomas commented:
Since this has been released into Kinetic as well as Jade for a while without any regressions I cherry-picked the change to the _indigo-devel_ branch as well: 2ce1584f0d96dbce6986781b56f79609c2492435

It will be part of the next patch release for Indigo.
